### DISCUSSION TITLE
NTP issue with 9.02 but not 9.00

### DISCUSSION THREAD

**Discussion by marioamiga** [astlouys/Pico-Green-Clock#28] [General]
Hi, firstly thanks for this work, its excellent.

I was running 9.02 with a Pico W.  The NTP element is key for my use, but today when the UK clocks changed, my clock did not change as expected.  Following a mornings investigating, I realised the NTP element was just not working at all, despite connecting to my router wifi just fine.  The LED on the rear of the clock also remained on.

Following various re-downloads to the clock, nothing worked wrt NTP operation.  In desperation, I tried 9.00 and hey presto, correct NTP operation and the LED on the rear was off.  If I try and manually change the time now, the NTP function corrects it back to the correct time.

I cannot see any other comments on Github related to this problem, so throught I'd post my issue and seek further comment.

Regards

Neil

**Comment by astlouys** [comment]
Hi Neil,

Thank you for your good words and for your feedback.

I know there has been changes in the Pico SDK update around the time I released Firmware Version 9.00 and I strongly suspect it impacted the NTP behavior in my code (see other thread on the subject). If ever you rebuild Version 9.02 (instead of using the .uf2 file in my repository, you may want to make sure that you have the latest SDK version.

Other than that. while working on the NTP feature, I had to update / change different parts of the code that could have had impacts on the automatic Summer Time / Winter Time handling (explaining why your clock didn't change automatically). I'll have to recheck this aspect, although fine tuning NTP may be the simplest solution.

As for the problems you encountered with your NTP connection, does your Network name (or Password) use special characters that have a special meaning in the C language ? (Like "Escape", "Back-slash", "Percent sign", etc... you said that Version 9.00 works fine, so I assume this is not the case since the problem would appear in 9.00 as well as 9.02).

Also, I'm aware of the following bug: if your Green Clock remains On while your WiFi network goes down, then back On, the Green Clock will not properly re-sync with your network once it is restored. You will have to turn the clock Off, then back On to re-sync. I expect this bug will be simple to fix in the next release, but this is a bug in the current code. Not sure if this may explain some of your problems...

Let me know if ever you find anything else...

**Reply by oldmaker** [reply]
Hi everyone,
for documentation only: I'm running 9.02 with a Pico W, too, and the time changed perfectly.
I am sure I used the latest Pico SDK version to rebuild the firmware, when I added the italian translation.
Pico Green Clock with André's firmware works really well; I can't say the same thing about my radio-controlled clocks: some of them did not change the time, sadly: the DCF77 signal comes from Frankfurt and nowadays that is too far a distance to have a good radio signal.
Bye

**Comment by marioamiga** [comment]
So I've tried several times with a Pi server and windows based solutions to compile your 9.0.3 C file, but without success.  I keep running into various problems and I never get to a final .uf2 file.  I will be trying again soon though to give it another go as I would like to be running your latest version.  Thanks again for the great code...bye for now

**Comment by astlouys** [comment]
Hi Neil,

You don't give much details about your problem, so it's difficult for me to help...

1) Are you using the ".uf2" file that I put on GitHub, or are you building your own executable ?
2) If you're building your own version, are you using the latest SDK version ?
3) Are you using Visual Studio Code for your development ?
4) Does your compiler / linker give you any error / warning message ?
5) Which compiler did you select ?
6) Have you connected a terminal emulator to the USB port of the Pico to see if there are some error messages (although most messages are turned Off in a release version).
7) Did you make sure that your WiFi network was On (and remained On all the time) BEFORE starting the Green Clock ?
8) Is there any "strange" character in your network name / password that has a specific meaning in the C language (percent sign, slash, back slash, etc) ?
9) Have you been able to build an executable (".uf2" file) and run it on the Pico even if the NTP feature does not work ?
10) What is the complete behaviour of the green LED on the Pico from start and up to 90 seconds later ?  Is it blinking many times ?  How many blinks and how many times ?

I must say that all my development is done on a Raspberry Pi 4 (waiting for the Pi 5 to be available), so if there is something special to know about a Windows environment, I may not be of great help...

I'll try to help with the answers you give. If other users have an idea, they are welcome to join us...

**Comment by marioamiga** [comment]
Hi, thanks for the reply. I'll try to answer each point the best I can.

    Are you using the ".uf2" file that I put on GitHub, or are you building your own executable ? - YES, bothe the 9 and 9.02 versions

    If you're building your own version, are you using the latest SDK version ? - I always download the latest when I attempt to compile

    Are you using Visual Studio Code for your development ?  I've tried that and building a server with a Pi 4, no success yet

    Does your compiler / linker give you any error / warning message ?It did, but sorry I can't remember.  I will note it down the next time

    Which compiler did you select ? I've followed various tutorials for both platforms, again I'll note specifically the next time

    Have you connected a terminal emulator to the USB port of the Pico to see if there are some error messages (although most messages are turned Off in a release version). - NO

    Did you make sure that your WiFi network was On (and remained On all the time) BEFORE starting the Green Clock ? - YES

    Is there any "strange" character in your network name / password that has a specific meaning in the C language (percent sign, slash, back slash, etc) ? - Nope, numbers and letters only

    Have you been able to build an executable (".uf2" file) and run it on the Pico even if the NTP feature does not work ? - Can't get to a .uf2 thus far

    What is the complete behaviour of the green LED on the Pico from start and up to 90 seconds later ? Is it blinking many times ? How many blinks and how many times ? - Sorry, did not note.

I will try again in the next few days and report back

Neil

**Comment by gatheall** [comment]
For what it's worth, I ran into a couple of problems when compiling the source for 9.03 back in September on a rpi 3.  I don't know if they affect you, Neil, but I'll mention them in case you or others run into them. I'll also add that this was my first time building something for the Pico.

- I was getting errors about pico_sdk_import.cmake not being found even when PICO_SDK_PATH was set and the file existed there.  I found   some generic instructions suggesting the need to copy the file into your project directory; eg, while in Pico-Green-Clock :

>     cp -i ../pico-sdk/external/pico_sdk_import.cmake .

- I ran into errors when trying to run cmake saying : 

>   CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
>   Missing variable is:
>    CMAKE_ASM_COMPILE_OBJECT

  Googling turned up a post on sourceforge suggesting the fix is to edit the CMakeLists.txt and add "ASM C CXX" to the project() call and then run cmake again.

George

**Reply by astlouys** [reply]
Thank you for your feedback George. I hope it could help Neil with the problems he encounter.

I would add something that I found useful:
Instead of copying the "pico_sdk_import.cmake" file into the project directory, generate a symbolic link (into every project directory) to the "official" version of the file:  "ln  -s  /home/pi/pico/pico-sdk/external/pico_sdk_import.cmake" (without the double quotes).

I used the usual path in my example above, adapt it if needed.

By doing so, you're sure you will always use the latest version of the file (assuming you take the time to update it on your system) and you won't have to update all you project directories every time.

Since I use Visual Studio Code (VSC) for my development, many actions are automatically taken by VSC "behind the scene".
